File: program/program/implement-jump-search/implement_jump_search.hs
import Data.List

-- Jump search function
jumpSearch :: (Ord a) => [a] -> a -> Maybe Int
jumpSearch arr x
    | null arr = Nothing
    | otherwise =
        let n = length arr
            blockSize = floor (sqrt (fromIntegral n))
            recJumpSearch prev blockSize
                | prev >= n = Nothing
                | arr !! prev < x =
                    recJumpSearch (min (prev + blockSize) (n - 1)) blockSize
                | arr !! prev == x = Just prev
                | otherwise =
                    let recLinearSearch idx
                            | idx >= min (prev + blockSize) (n - 1) = Nothing
                            | arr !! idx == x = Just idx
                            | otherwise = recLinearSearch (idx + 1)
                    in recLinearSearch prev
        in recJumpSearch 0 blockSize

-- Function to read a list of integers from the user
readIntList :: IO [Int]
readIntList = do
    putStrLn "Enter a list of integers separated by spaces:"
    input <- getLine
    let parsedInput = map read $ words input :: [Int]
    return parsedInput

main :: IO ()
main = do
    -- Read the list of integers from the user
    arr <- readIntList

    -- Sort the list
    let sortedArr = sort arr

    -- Read the value to be searched from the user
    putStrLn "Enter the value to search for:"
    value <- readLn :: IO Int

    -- Perform jump search and print the result
    case jumpSearch sortedArr value of
        Just idx -> putStrLn $ "Value found at index: " ++ show idx
        Nothing -> putStrLn "Value not found in the list."

File: program/program/implement-jump-search/ImplementJumpSearch.c
#include <stdio.h>
#include <math.h>

int jumpSearch(int arr[], int n, int x) {
    int blockSize = sqrt(n);
    int prev = 0;

    while (arr[(int)(fmin(blockSize, n) -1 )] < x) {
        prev = blockSize;
        blockSize += sqrt(n);
        if (prev >= n)
            return -1; 
    }

    while (arr[prev] < x) {
        prev++;

        if (prev == fmin(blockSize, n))
            return -1; 
    }

    if (arr[prev] == x)
        return prev; 

    return -1; 
}

int main() {
    int arr[] = {1, 2, 3, 4, 5};
    int n = sizeof(arr) / sizeof(arr[0]);

    int x = 5;
    int result = jumpSearch(arr, n, x);

    if (result != -1)
        printf("Output: %d\n", result);
    else
        printf("Value not found in the list\n");

    return 0;
}
